cmd/clusterctl/phases/pivot_test.go

```diff
@@ -0,0 +1,326 @@
+
+func (p *providerComponents) String() string {
+	yamls := []string{}
+	for _, name := range p.names {
+		yamls = append(yamls, fmt.Sprintf(`kind: StatefulSet
+apiVersion: apps/v1
+metadata:
+    name: %s
+`, name))
+	}
+	return strings.Join(yamls, "---")
+}
+
+func TestPivot(t *testing.T) {
+	pc := &providerComponents{
+		names: []string{"test1", "test2"},
+	}
+
+	namespace := "default"
+
+	source := NewSourcer().
+		WithCluster(namespace, "my-cluster").
+		WithMachineDeployment(namespace, "my-machine-deployment").
+		WithOwnedMachineSet(namespace, "my-machine-deployment").
+		WithMachines(namespace, "my-cluster")
+
+	expectedClusters := len(source.clusters[namespace])
+	expectedMachineDeployments := len(source.machineDeployments[namespace])
+	expectedMachineSets := len(source.machineSets[namespace])
+	expectedMachines := len(source.machines[namespace])
+
+	target := &target{
+		clusters:           make(map[string][]*clusterv1.Cluster),
+		machineDeployments: make(map[string][]*clusterv1.MachineDeployment),
+		machineSets:        make(map[string][]*clusterv1.MachineSet),
+		machines:           make(map[string][]*clusterv1.Machine),
+	}
+
+	if err := Pivot(source, target, pc.String()); err != nil {
+		t.Fatalf("did not expect err but got %v", err)
+	}
+
+	if len(source.clusters[namespace]) != 0 {
+		t.Fatal("should have deleted all capi clusters from the source k8s cluster")
+	}
+	if len(source.machineDeployments[namespace]) != 0 {
+		t.Fatal("should have deleted all machine deployments from the source k8s cluster")
+	}
+	if len(source.machineSets[namespace]) != 0 {
+		t.Fatal("should have deleted all machine sets from source k8s cluster")
+	}
+	if len(source.machines[namespace]) != 0 {
+		t.Fatal("should have deleted all machines from source k8s cluster")
+	}
+
+	if len(target.clusters[namespace]) != expectedClusters {
+		t.Fatal("expected cluster to pivot")
+	}
+	if len(target.machineDeployments[namespace]) != expectedMachineDeployments {
+		t.Fatal("expected machinedeployments for cluster to pivot")
+	}
+	if len(target.machineSets[namespace]) != expectedMachineSets {
+		t.Fatal("expected machines sets to pivot")
+	}
+	if len(target.machines[namespace]) != expectedMachines {
+		t.Fatal("expected machines to pivot")
+	}
+}
+
+func NewSourcer() *sourcer {
+	return &sourcer{
+		clusters:           make(map[string][]*clusterv1.Cluster),
+		machineDeployments: make(map[string][]*clusterv1.MachineDeployment),
+		machineSets:        make(map[string][]*clusterv1.MachineSet),
+		machines:           make(map[string][]*clusterv1.Machine),
+		count:              0,
+	}
+}
+func (s *sourcer) WithCluster(ns, name string) *sourcer {
+	s.clusters[ns] = append(s.clusters[ns], &clusterv1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: ns,
+		},
+	})
+	return s
+}
+
+// Associations are done via namespaces. If you create a machinedeployment in
+// the same namespace as a cluster then CAPI will assume it owns that machinedeployment
```

Comment:
The association via namespace doesn't hold true anymore (at least for the general codebase). I need to rebase to pick up the changes that include Vince's MachineSet/MachineDeployment ownerRef updates.

Association is now done both by Namespace and the same label association as is done for Machines